### ISSUE TITLE
TypeScript doesn't warn me that a property is missing

### ISSUE THREAD

**Issue by asdf1899** [microsoft/TypeScript#50845]
# Bug Report

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.2#code/C4TwDgpgBAIghsOAVc0C8UDeAoKUBucANgK4QBcUAFPIlAGRQCiAHsAE5wCUUAPrAjgBubAF8R2UJAF0MOPFIgBGSgGcOASwB2AcxHjsk1Mzacoc3FA2qmAEw2IARkQpRHAe3cu4WkXndaAMIAFj46rlQ8aAB8BO4atvoSAMYB6lC2gpS0yMYWeAD0BQTEZFCqwe4kRLZQwMHs7gDuUD5QEOyN7NQBIWHQ1lBa7sDlkMkaAGYaELZcloSkrvJ4dagqUADkipsANJZ41nYOcM6uHGT7eKL74kA)

### 💻 Code

```ts
type DataType = {
  value: (Data & Extra) | Data;
};

type Data = {
  type1: string;
};

type Extra = {
  isEditable: boolean;
  onChange: () => void;
};

const data: DataType = {
  // value should throw an error (onChange is not specified)
  value: {
    type1: 'type',
    isEditable: true,
  },
};
```

### 🙁 Actual behavior

TypeScript ignores this definition (`(Data & Extra) | Data`)

### 🙂 Expected behavior

 I expect TypeScript to warn me that a property within `value` is missing: e.g (`onChange` is not specified)

**Comment by fatcerberus** [comment]
The value you've provided is not a valid `Data & Extra`, but it *is* a valid `Data`.  Since the type of `.value` is `(Data & Extra) | Data`, there is no type error in this code.

**Comment by fatcerberus** [comment]
Just to reiterate: Because objects can have more properties than their type calls for, unions of object types are not mutually exclusive.  If you want a mutually exclusive type you need to use a discriminated union.

**Comment by MartinJohns** [comment]
Also related: #20863
